content/buildah-podman-skopeo.howto.md

```diff
@@ -0,0 +1,238 @@
+---
+title: "Building Containers safely without Docker (Look Ma! No Docker!)"
+date: 2020-09-29T13:42:36-07:00 
+draft: true
+---
+
+If you want to create a container, the traditional way to go about it is to use just `docker build .` and let docker take care of it for you.
+
+One problem with this is that it requires access to the Docker daemon which then runs as root. It also tends to lead to more bloated images as the typical Dockerfile starts with pulling in someone elses' base container (not always bad, just be aware of where your base containers come from).
+
+These problems become more important when you start automating container builds as someone could use a CI/CD pipeline to execute their own code which would get run with elevated privileges! Removing the dependency on the Daemon removes this path of exploits.
+
+
+My preferred way to build my container images is with [Buildah](https://buildah.io)!
+
+Note: I'll probably be using "container" and "container image" interchangeably. It should be obvious from context what I mean.
+
+Buildah uses the standard CRI set of tooling and configuration which means that it runs very nicely along side CRI-O and Podman making for a full (mostly drop in) replacement for Docker.
+
+# What is a Container (simply)
+
+A container is a collection of filesystem layers that when joined together provide everything needed to run the application in a self-contained way. It also includes a manifest that describes how to use what’s inside the container.
+The process for building a container is to build a layer, save it, build the next layer, save it, etc... until you have everything you want in there. The reason for splitting out layers is to ease both the building and splitting out of layers. 
+If a certain part of your build step always builds the same thing, why repeat that work? The same is for transport, if there’s a large baseline that each container needs, you don’t need to download it every time.
+
+
+[Some graphic here?](!/img/asdas )
+
+As these are really just a collection of bits, there's no magic to building one that requires a centralized service to do the build. Also, removing the 
+
+
+# Toolset
+
+There a little bit of overlap between the tools but they each do a different part of the puzzle.]
+
+Buildah builds OCI compliant images. It can allow you to run commands in the container’s context as well as mount the container’s filesystem to move files in from elsewhere 
+
+Podman runs images in a CRI compatible runtime (such as CRI-O). It is a drop in replacement for Docker and supports all of the same subcommands.
```

Comment:
IIRC you could simply do a "podman build" too?

Reply:
You could but there are a lot of build pipelines where a Dockerfile might not be the right choice